### PR TITLE
Feature/rb 110 function receive new enemy

### PR DESCRIPTION
**Pull request by guillaumeAbel**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] Norm of the code has been respected

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of change does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [ ] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **What is the current behavior?** (link an issue based on the kind of change this pr introduce)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:

### Summary by CodeRabbit

- New Feature: Added network system manager to enhance multiplayer capabilities.
- New Feature: Introduced parallax effect management for improved visual experience.
- Improvement: Enhanced enemy initialization with additional parameters for better game dynamics.
- Improvement: Simplified code and improved performance by direct access to enemy type.
- Improvement: Added logging statements across various functions for better debugging and system tracking.
- Improvement: Implemented thread safety measures in the client network system to prevent data races.
- Improvement: Modified enemy struct for clearer and more consistent code.
- Improvement: Added a function to directly access the last element in an array, improving code efficiency.
- Improvement: Updated player and missile creation code for better game mechanics.

### CODE REVIEW

**Review by kw4ntiq — APPROVED** [review]

File: src/Client/Systems/Network/ClientNetwork.cpp [hunk 1]
```diff
@@ -67,8 +67,16 @@ namespace Systems {
...
         const auto newEnemy = std::any_cast<struct msgNewEnemy_s>(any);
-        initEnemy(enemyFile);
+        reviveEnemy(newEnemy.enemyInfos.id);
+        Types::Position pos = {
+            static_cast<float>(newEnemy.enemyInfos.pos.x),
+            static_cast<float>(newEnemy.enemyInfos.pos.y)};
+        struct health_s hp = newEnemy.enemyInfos.life;
+        Registry::getInstance().getComponents<Types::Position>().insertBack(pos);
+        Registry::getInstance().getComponents<struct health_s>().insertBack(hp);
+        Registry::getInstance().getComponents<Types::Enemy>().back().constId = newEnemy.enemyInfos.id;
```
Comment (romainpanno) [inline src/Client/Systems/Network/ClientNetwork.cpp]: delete this line

**Review by romainpanno — CHANGES_REQUESTED** [review]

**Review by romainpanno — APPROVED** [review]

**Review by Saverio976 — APPROVED** [review]

File: src/ECS/Systems/Systems.cpp [hunk 2]
```diff
@@ -191,37 +192,35 @@ namespace Systems {
...
-            nlohmann::basic_json<> animRectData = Json::getInstance().getDataFromJson(elem, "animRect");
-            Types::AnimRect animRect            = {
-                rect,
-                Json::getInstance().getDataFromJson(animRectData, "move").get<std::vector<Types::Rect>>(),
-                Json::getInstance().getDataFromJson(animRectData, "attack").get<std::vector<Types::Rect>>(),
-                Json::getInstance().getDataFromJson(animRectData, "dead").get<std::vector<Types::Rect>>()};
-
 #ifdef CLIENT
             Registry::getInstance().getComponents<Raylib::Sprite>().insertBack(enemy);
             Registry::getInstance().setToFrontLayers(id);
```
Comment (romainpanno) [inline src/ECS/Systems/Systems.cpp:219]: remove this

File: src/ECS/Systems/Systems.cpp [hunk 3]
```diff
@@ -191,37 +192,35 @@ namespace Systems {
...
 
-            nlohmann::basic_json<> animRectData = Json::getInstance().getDataFromJson(elem, "animRect");
-            Types::AnimRect animRect            = {
-                rect,
-                Json::getInstance().getDataFromJson(animRectData, "move").get<std::vector<Types::Rect>>(),
-                Json::getInstance().getDataFromJson(animRectData, "attack").get<std::vector<Types::Rect>>(),
-                Json::getInstance().getDataFromJson(animRectData, "dead").get<std::vector<Types::Rect>>()};
-
 #ifdef CLIENT
             Registry::getInstance().getComponents<Raylib::Sprite>().insertBack(enemy);
```
Comment (romainpanno) [inline src/ECS/Systems/Systems.cpp]: change bien spritedata

File: src/ECS/Systems/Systems.cpp [hunk 4]
```diff
@@ -191,37 +192,35 @@ namespace Systems {
...
+                rect,
+                Json::getInstance().getDataFromJson(animRectData, "move").get<std::vector<Types::Rect>>(),
+                Json::getInstance().getDataFromJson(animRectData, "attack").get<std::vector<Types::Rect>>(),
+                Json::getInstance().getDataFromJson(animRectData, "dead").get<std::vector<Types::Rect>>()};
 #else
             Registry::getInstance().addEntity();
-
 #endif
-            Types::Enemy enemyComp   = {};
+            Types::Enemy enemyComp   = {enemyId};
```
Comment (romainpanno) [inline src/ECS/Systems/Systems.cpp]: sprite data

**Review by romainpanno — CHANGES_REQUESTED** [review]

File: src/ECS/SparseArray.hpp [hunk 5]
```diff
@@ -65,6 +65,14 @@ class SparseArray {
...
+        {
+            if (_dense.empty()) {
+                throw std::runtime_error("SparseArrays::back: empty!");
+            }
+            return _dense.back();
+        }
+
         /*
          * A dense sparseArrays is not sort by entities id, the begin of two
          * sparseArrays could be different entities, only _sparse are
```
[the remark this answers is not in the document]
Comment (guillaumeAbel) [reply src/ECS/SparseArray.hpp:78]: lit la ligne d'apres fdp

File: src/ECS/ECSCustomTypes.hpp [hunk 6]
```diff
@@ -131,10 +130,11 @@ namespace Types {
                 return _enemyNb;
             }
 
+            enemy_id_s constId;
+            enum enemy_type_e type;
```
[the remark this answers is not in the document]
Comment (guillaumeAbel) [reply src/ECS/ECSCustomTypes.hpp:134]: yeah bro its struct

**Review by romainpanno — APPROVED** [review]

**Review by Saverio976 — APPROVED** [review]

**Review by romainpanno — APPROVED** [review]

**Pull request merged**